src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,199 @@
+            
+            // this method performs a deferred registration: the type name specified is checked when the factory is requested for the first time. 
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factoryTypeAssemblyQualifiedName, null);
+        }
+        
+        public static void RegisterFactory(string providerInvariantName, Type providerFactoryClass)
+        {
+            RegisterFactory(providerInvariantName, GetFactoryInstance(providerFactoryClass));
+        }
+
+        public static void RegisterFactory(string providerInvariantName, DbProviderFactory factory)
+        {
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            ADP.CheckArgumentNull(factory, nameof(factory));
+
+            _registeredFactories[providerInvariantName] = new ProviderRegistration(factory.GetType().AssemblyQualifiedName, factory);
+        }
+        
+        public static bool UnregisterFactory(string providerInvariantName)
+        {
+            return !string.IsNullOrWhiteSpace(providerInvariantName) && _registeredFactories.TryRemove(providerInvariantName, out ProviderRegistration providerRegistration);
+        }
+        
+        private static DbProviderFactory GetFactory(string providerInvariantName, bool throwOnError)
+        {
+            if (throwOnError)
+            {
+                ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+            }
+            else
+            {
+                if (string.IsNullOrWhiteSpace(providerInvariantName))
+                {
+                    return null;
+                }
+            }
+            bool wasRegistered = _registeredFactories.TryGetValue(providerInvariantName, out ProviderRegistration registration);
+            if (!wasRegistered)
+            {
+                return throwOnError ? throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_InvariantNameNotFound, providerInvariantName)) : (DbProviderFactory)null;
+            }
+            DbProviderFactory toReturn = registration.FactoryInstance;
+            if (toReturn == null)
+            {
+                // Deferred registration, do checks now on the type specified and register instance in storage.
+                // Even in the case of throwOnError being false, this will throw when an exception occurs checking the registered type as the user has to be notified the 
+                // registration is invalid, even though the registration is there.
+                Type providerType = Type.GetType(registration.FactoryTypeAssemblyQualifiedName);
+                if (null == providerType)
+                {
+                    throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_FactoryNotLoadable, registration.FactoryTypeAssemblyQualifiedName));
+                }
+                toReturn = GetFactoryInstance(providerType);
```

Comment:
Nit: It seems the 6 lines above are repeated in GetFactory(providerRow), can extract another private method?

Reply:
It's a clone indeed. I'll go with GetFactoryInstanceFromTypeName(aqn) and it will cover these lines, and always return a value or throw the exception mentioned above.